libsolidity/codegen/CompilerUtils.cpp

```diff
@@ -317,53 +317,65 @@ void CompilerUtils::memoryCopyPrecompile()
 	m_context << Instruction::POP << Instruction::POP << Instruction::POP;
 }
 
-void CompilerUtils::memoryCopy32()
+void CompilerUtils::memoryCopy(bool _copyFullWords)
 {
-	// Stack here: size target source
-
-	m_context.appendInlineAssembly(R"(
+	// Stack here: len dst src
+	string wordCopyCode = R"(
+		jumpi(skipwordcopy, eq(len, 0))
 		{
-		jumpi(end, eq(len, 0))
-		start:
-		mstore(dst, mload(src))
-		jumpi(end, iszero(gt(len, 32)))
-		dst := add(dst, 32)
-		src := add(src, 32)
-		len := sub(len, 32)
-		jump(start)
-		end:
+			let last := add(src, and(sub(len, 1), not(31)))
+			let delta := 32
+			jumpi(startwordcopy, or(gt(src, dst), lt(add(last, 31), dst)))
+			// We have to copy in reverse to avoid overwriting
+			delta := sub(0, delta)
+			dst := add(dst, sub(last, src))
+			{
+				let t := src
+				src := last
+				last := t
+			}
+			startwordcopy:
```

Comment:
In the reverse case, either
* the partial copy should happen before the word copy, or
* the partial snippet should be made immediately after `src`, not immediately before `dst`.